include/zephyr/sys/reboot.h

```diff
@@ -20,19 +13,48 @@
 extern "C" {
 #endif
 
-#define SYS_REBOOT_WARM 0
-#define SYS_REBOOT_COLD 1
+/**
+ * @brief System reboot
+ * @defgroup sys_reboot System reboot
+ * @ingroup os_services
+ * @{
+ */
+
+/** Reboot types. */
+enum sys_reboot_type {
+	/** Cold reboot (no state is preserved). */
+	SYS_REBOOT_COLD,
+	/** Warm reboot. */
+	SYS_REBOOT_WARM,
+};
```

Comment:
Not trying to open an endless discussion, but: 
> Cold reboot is basically what you expect to happen on powering up a device;

Agreed.

> warm reset may be..

..a lot of different things, and for HW assisted reboots triggered by SW, the HW may provide several options with significantly different outcomes.
Which makes me wonder if Zephyr should not make it easier for platforms to define their own different reboot types. This is probably out of scope for this PR, though this enum seems to be tying things up?

Reply:
I think it's easy to agree on what cold boot means, but not so much on "warm". I'm not even sure we're now able to support something like warm reset properly, see, for example, https://trustedfirmware-a.readthedocs.io/en/latest/design/reset-design.html

reboot types have been undocumented from day 0 https://github.com/zephyrproject-rtos/zephyr/commit/826edfc16b2615e8ee5f8fd0b06af39ded365a8e, who knows what they meant in 2016...